Cleanup 3-to-4 renames, prevent common words replacements, handle project.godot settings

akien-mga opened the pull request:
A bit late but should help solve common pitfalls.
I'll put some inline comments on some relevant changes.

- ~Fixes _not_ #66061.~
- Fixes #73505.
- Fixes #73996.
- Fixes #66125.
- Supersedes #66251.

akien-mga committed:
Cleanup 3-to-4 renames, prevent common words replacements

Fixes #73505.
Fixes #73996.

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -456,7 +466,6 @@ const char *RenamesMap3To4::gdscript_function_renames[][2] = {
 	{ "post_import", "_post_import" }, // EditorScenePostImport
 	{ "print_stray_nodes", "print_orphan_nodes" }, // Node
 	{ "property_list_changed_notify", "notify_property_list_changed" }, // Object
-	{ "raise", "move_to_front" }, // CanvasItem
```
Removed as possibly conflicting with comments / user-defined methods.

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -570,26 +581,21 @@ const char *RenamesMap3To4::gdscript_function_renames[][2] = {
 	{ "clamped", "limit_length" }, // Vector2
 	{ "get_rotation_quat", "get_rotation_quaternion" }, // Basis
 	{ "grow_margin", "grow_side" }, // Rect2
-	{ "invert", "reverse" }, // Array -- Give it a check. Be careful, this will be used everywhere.
```
Too common.

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -570,26 +581,21 @@ const char *RenamesMap3To4::gdscript_function_renames[][2] = {
 	{ "find_last", "rfind" }, // Array, String
 	{ "to_ascii", "to_ascii_buffer" }, // String
 	{ "to_utf8", "to_utf8_buffer" }, // String
-	{ "to_wchar", "to_utf32_buffer" }, // String -- utf32 or utf16?
```
Wrong (will be fixed by https://github.com/godotengine/godot/pull/73225).

YuriSizov replied:
So, shouldn't it be changed to convert to `to_wchar_buffer` then?

akien-mga replied:
It's not merged yet, because production folks are stringy on their merges :P

YuriSizov replied:
Ahh, right right, the vermin of this world get in our way again!

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -154,6 +154,12 @@ const char *RenamesMap3To4::enum_renames[][2] = {
 };
 
 const char *RenamesMap3To4::gdscript_function_renames[][2] = {
+	// NOTE: Commented out renames are disabled because deemed not suitable for
+	// the current way the regex-based converter works.
+	// When uncommenting any of those as suitable for conversion, please move it
+	// to the block with other enabled conversions, ordered alphabetically, and
+	// make sure to add it to the C# rename map too.
```
Added this blurb so we don't have to keep commented out stuff in sync too.

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -617,52 +623,6 @@ const char *RenamesMap3To4::gdscript_function_renames[][2] = {
 
 // gdscript_function_renames clone with CamelCase
 const char *RenamesMap3To4::csharp_function_renames[][2] = {
```
For properties I did a full comparative review of GDScript vs C#, but not for this one.

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -1060,40 +1016,49 @@ const char *RenamesMap3To4::csharp_function_renames[][2] = {
+
 	{ "as_normalmap", "as_normal_map" }, // NoiseTexture
 	{ "bbcode_text", "text" }, // RichTextLabel
-	{ "bg", "panel" }, // Theme
```
`bg`, `ofs`, `on`, `off` all seem too common.

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -1060,40 +1016,49 @@ const char *RenamesMap3To4::csharp_function_renames[][2] = {
 	{ "caret_blink_speed", "caret_blink_interval" }, // TextEdit, LineEdit
 	{ "caret_moving_by_right_click", "caret_move_on_right_click" }, // TextEdit
 	{ "caret_position", "caret_column" }, // LineEdit
+	{ "cast_to", "target_position" }, // RayCast2D, RayCast3D
```
Seems safe enough? There's `Object::cast_to` but it's not part of the scripting API.

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -1107,6 +1072,7 @@ const char *RenamesMap3To4::gdscript_properties_renames[][2] = {
 	{ "drag_margin_top", "drag_top_margin" }, // Camera2D
 	{ "drag_margin_v_enabled", "drag_vertical_enabled" }, // Camera2D
 	{ "enabled_focus_mode", "focus_mode" }, // BaseButton - Removed
+	{ "extents", "size" }, // BoxShape3D, LightmapGI, ReflectionProbe
```
Seems fine now that we ported ReflectionProbe to `size`.

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -1155,6 +1118,7 @@ const char *RenamesMap3To4::gdscript_properties_renames[][2] = {
 	{ "refuse_new_network_connections", "refuse_new_connections" }, // MultiplayerAPI
 	{ "region_filter_clip", "region_filter_clip_enabled" }, // Sprite2D
 	{ "reverb_bus_enable", "reverb_bus_enabled" }, // Area3D
+	{ "scancode", "keycode" }, // InputEventKey
```
Seems safe enough?

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -1170,49 +1134,29 @@ const char *RenamesMap3To4::gdscript_properties_renames[][2] = {
 	{ "table_hseparation", "table_h_separation" }, // Theme
 	{ "table_vseparation", "table_v_separation" }, // Theme
-	{ "translation", "position" }, // Node3D -- Breaks GLTFNode
+	{ "toplevel", "top_level" }, // Node
```
Seems safe enough?

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -1221,13 +1165,17 @@ const char *RenamesMap3To4::csharp_properties_renames[][2] = {
 	{ "FocusNeighbourRight", "FocusNeighborRight" }, // Control
 	{ "FocusNeighbourTop", "FocusNeighborTop" }, // Control
 	{ "FollowViewportEnable", "FollowViewportEnabled" }, // CanvasItem
+	{ "FileIconModulate", "FileIconColor" }, // Theme
```
I don't know if theme properties are actually capitalized or if they're used with their core snake_case identifier.
Either way, these shouldn't hurt as it will do nothing if they're not capitalized.

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -1424,15 +1378,16 @@ const char *RenamesMap3To4::shaders_renames[][2] = {
+	// { "Reference", "RefCounted" },
+	// { "Tabs", "TabBar" },
+	// { "Shape", "Shape3D" },
+	// { "Path", "Path3D" },
```
This will lead to more manual work for users, as Reference, Area, Camera, Shape and Path are quite used. But those are also potential clashes with user comments and strings.

akien-mga commented on editor/renames_map_3_to_4.cpp:
```diff
@@ -1633,19 +1575,11 @@ const char *RenamesMap3To4::class_renames[][2] = {
 	{ "XROrigin", "XROrigin3D" },
 	{ "YSort", "Node2D" }, // CanvasItem has a new "y_sort_enabled" property.
 
-	// Portal and room occlusion culling was replaced by raster occlusion culling.
-	{ "CullInstance", "Node3D" },
-	{ "RoomGroup", "Node3D" },
-	{ "Room", "Node3D" },
-	{ "RoomManager", "Node3D" },
-	{ "Portal", "Node3D" },
```
This kind of renames was added as aliases in `register_scene_types.cpp` to avoid scenes getting corrupted, but it makes no sense to do the same in scripts.

YuriSizov commented:
Removing Reference is a big downgrade, IMO. There is a decent possibility for conflicts, but it's also a core class in Godot, and can be used quite a lot. Not sure if it's better to avoid converting it and have users update it manually.

akien-mga commented:
That's fair. What about the others?
```
	// { "Area", "Area3D" },
	// { "Camera", "Camera3D" },
	// { "Particles", "GPUParticles3D" },
	// { "Path", "Path3D" },
	// { "Reference", "RefCounted" },
	// { "Shape", "Shape3D" },
	// { "Tabs", "TabBar" },
	// { "World", "World3D" },
```

I think at least "World" can be kept commented out, currently it would replace stuff like using `World` as node name in a node path (`$Game/World/0/StartPosition`, etc.).

"Particles" probably doesn't make sense to convert like this because it's not going to be compatible anyway.

The rest might be still worth it even if there's a high risk of doing wrong replacements in e.g. comments, strings and node paths?

YuriSizov reviewed (approved):
Aside from the two notes above, LGTM. We can give it a go.

> The rest might be still worth it even if there's a high risk of doing wrong replacements in e.g. comments, strings and node paths?

I agree that World is too commonplace (I have it in practically every project, and it's never the `World` node). For the rest, it's a toss. We can go on with a safer option and re-evaluate later. I wouldn't worry too much about comments though.

akien-mga commented:
> That's fair. What about the others?
> 
> ```
> 	// { "Area", "Area3D" },
> 	// { "Camera", "Camera3D" },
> 	// { "Particles", "GPUParticles3D" },
> 	// { "Path", "Path3D" },
> 	// { "Reference", "RefCounted" },
> 	// { "Shape", "Shape3D" },
> 	// { "Tabs", "TabBar" },
> 	// { "World", "World3D" },
> ```

Alright, re-added those aside from "World" (too common) and "Particles" (too common, and incompatible class).

YuriSizov commented:
Feels good!

akien-mga committed:
Add 3-to-4 renames for project settings in project.godot

In the ConfigFile format, the first subpath is the category and is not part
of the line that the regex would match.

Fixes #66125.

akien-mga commented:
Added a second commit to fix #66125 too, with a much simpler approach than #66251.
